### ISSUE TITLE
MySQL defaults to utf8mb4_bin collation instead of utf8mb4_unicode_ci

### ISSUE THREAD

**Issue by micaherne** [moodlehq/moodle-ci-runner#123]
The Moodle documentation for MySQL [here ](https://docs.moodle.org/404/en/MySQL#Configure_full_UTF-8_support) and [here ](https://docs.moodle.org/404/en/MySQL_full_unicode_support) say that MySQL should be configured with the utf8mb4_unicode_ci collation but moodle-ci-runner defaults to utf8mb4_bin which can cause issues.

I'm not sure if this is a bug or not as it seems to be consistently implemented in all the my.cnf files and the docker-database.sh script.

Should this maybe be changed to utf8mb4_unicode_ci?

To reproduce the issue you can install report_advanced grading on Moodle 4.3.3 and run moodle-ci-runner with the following config:

```
CODEDIR=/path/to/moodle
JOBTYPE=phpunit
PHP_VERSION=8.1
PHPUNIT_TESTSUITE=report_advancedgrading_testsuite
PHPUNIT_FILTER=test_guide
DBTYPE=mysqli
DBTAG=8.2
```

The test will fail, but will pass if you also set `DBCOLLATION=utf8mb4_unicode_ci`. (I don't claim to understand why this fails but it does.)

**Comment by andrewnicols** [comment]
Is this the failure you see: https://docs.moodle.org/dev/Common_unit_test_problems#Specific_failures

**Comment by micaherne** [comment]
No, it's this:

```
report_advancedgrading\locallib_test::test_guide
Failed asserting that an array contains 'admin_452525f043d9bb60aa3da1ebf45ca6b5'.

/var/www/html/report/advancedgrading/tests/locallib_test.php:235
/var/www/html/lib/phpunit/classes/advanced_testcase.php:81
```

It's running an SQL query which has `ORDER BY lastname ASC, firstname ASC etc` where there are non-ASCII names in the user table. The utf8mb4_bin collation is causing it to come back in a different order than would be expected on a production system with the correct collation (and PostgreSQL).

**Comment by stronk7** [comment]
Yeah, I'm afraid but we need to run unit tests with a case/accent sensitive collation to be able to cover everything.

Historically, the only case/accent sensitive collation available for MySQL has been the binary one. Luckily, since MySQL 8, they finally created case/accent sensitive collations for a lot of langs (apart from the general one).

So, in your case, I imagine that you may change to some of the available ones (`uft8mb4_xxxxx_as_cs`) and then you will get proper (sensitive) sorting for the "xxxxx" language. Or use the `uft8mb4_0900_as_cs` one, that is the "generic" one, not tidied to any language.

Disclaimer, I've not used any of those sensitive collations ever, I continue using the old "bin" one everywhere, but any of them (language-specific or no), should work.

Reference: https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-sets.html

Ciao :-)

**Comment by micaherne** [comment]
Thanks @stronk7, that's really useful! I did come across the new "as_cs" collations the other day when I was looking at this and they look promising. At the moment they cause some other unit test failures but I have a patch to add support for them, and I got the full unit test suite passing with it (including the ones in the link @andrewnicols sent where the recommended solution is to run them with utf8mb4_bin). When I get back to work on Tuesday I'll submit it through the tracker.
